internal/storage/webdav/webdav.go

```diff
@@ -0,0 +1,113 @@
+package webdav
+
+import (
+	"errors"
+	"io/fs"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/offen/docker-volume-backup/internal/storage"
+	"github.com/studio-b12/gowebdav"
+)
+
+type webDavStorage struct {
+	*storage.StorageBackend
+	client *gowebdav.Client
+	url    string
+}
+
+// NewStorageBackend creates and initializes a new WebDav storage backend.
+func NewStorageBackend(url string, remotePath string, username string, password string, urlInsecure bool,
+	logFunc storage.LogFuncDef) (storage.Backend, error) {
+
+	if username == "" || password == "" {
+		return nil, errors.New("newScript: WEBDAV_URL is defined, but no credentials were provided")
+	} else {
+		webdavClient := gowebdav.NewClient(url, username, password)
+
+		if urlInsecure {
+			defaultTransport, ok := http.DefaultTransport.(*http.Transport)
+			if !ok {
+				return nil, errors.New("newScript: unexpected error when asserting type for http.DefaultTransport")
+			}
+			webdavTransport := defaultTransport.Clone()
+			webdavTransport.TLSClientConfig.InsecureSkipVerify = urlInsecure
+			webdavClient.SetTransport(webdavTransport)
+		}
+
+		strgBackend := &storage.StorageBackend{
+			Backend:         &webDavStorage{},
+			DestinationPath: remotePath,
+			Log:             logFunc,
+			Name:            "WebDav",
+		}
+		webdavBackend := &webDavStorage{
+			StorageBackend: strgBackend,
+			client:         webdavClient,
+		}
+		strgBackend.Backend = webdavBackend
+		return strgBackend, nil
+	}
+}
+
+// GetName return the name of the storage backend through the interface
+func (stg *webDavStorage) GetName() string {
```

Comment:
Could we remove the needless indirection and define the value here once. Consumers can then just call the method instead of accessing the struct value, giving us a. a single way to access the name and b. prevent accidental mutation.

---

This applies to all packages.